VoodooPS2Trackpad/VoodooPS2ALPSGlidePoint.cpp

```diff
@@ -272,549 +415,3921 @@ bool ApplePS2ALPSGlidePoint::start( IOService * provider )
+         ((_packetByteCount == 4) && ((packet[3] & 0x48) != 0x48)) ||
+         ((_packetByteCount == 6) && ((packet[5] & 0x40) != 0x0)))) {
+        priv.PSMOUSE_BAD_DATA = true;
+        _ringBuffer.advanceHead(priv.pktsize);
+        return kPS2IR_packetReady;
+    }
+    
+    /* alps_is_valid_package_ss4_v2 */
+    if (priv.proto_version == ALPS_PROTO_V8 &&
+        ((_packetByteCount == 4 && ((packet[3] & 0x08) != 0x08)) ||
+         (_packetByteCount == 6 && ((packet[5] & 0x10) != 0x0)))) {
+        priv.PSMOUSE_BAD_DATA = true;
+        _ringBuffer.advanceHead(priv.pktsize);
+        return kPS2IR_packetReady;
+    }
+    
     packet[_packetByteCount++] = data;
-    if (kPacketLengthLarge == _packetByteCount ||
-        (kPacketLengthSmall == _packetByteCount && (packet[0] & 0xc8) == 0x08))
+    if (_packetByteCount == priv.pktsize)
     {
-        // complete 6 or 3-byte packet received...
-        _ringBuffer.advanceHead(kPacketLengthMax);
-        _packetByteCount = 0;
+        _ringBuffer.advanceHead(priv.pktsize);
         return kPS2IR_packetReady;
     }
     return kPS2IR_packetBuffering;
 }
 
-void ApplePS2ALPSGlidePoint::packetReady()
-{
+void ApplePS2ALPSGlidePoint::packetReady() {
     // empty the ring buffer, dispatching each packet...
-    while (_ringBuffer.count() >= kPacketLengthMax)
-    {
-        UInt8* packet = _ringBuffer.tail();
-        // now we have complete packet, either 6-byte or 3-byte
-        if ((packet[0] & 0xf8) == 0xf8)
-            dispatchAbsolutePointerEventWithPacket(packet, kPacketLengthLarge);
-        else
-            dispatchRelativePointerEventWithPacket(packet, kPacketLengthSmall);
-        _ringBuffer.advanceTail(kPacketLengthMax);
+    while (_ringBuffer.count() >= priv.pktsize) {
+        UInt8 *packet = _ringBuffer.tail();
+        if (priv.PSMOUSE_BAD_DATA == false) {
+            if (!ignoreall)
+                (this->*process_packet)(packet);
+        } else {
+            IOLog("ALPS: an invalid or bare packet has been dropped...\n");
+            /* Might need to perform a full HW reset here if we keep receiving bad packets (consecutively) */
+        }
+        _packetByteCount = 0;
+        _ringBuffer.advanceTail(priv.pktsize);
     }
 }
 
 // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 
-void ApplePS2ALPSGlidePoint::dispatchAbsolutePointerEventWithPacket(UInt8* packet, UInt32 packetSize)
-{
-    UInt32 buttons = 0;
-    int left = 0, right = 0, middle = 0;
-    int xdiff, ydiff, scroll;
-    uint64_t now_abs;
-    bool wasNotScrolling, willScroll;
+bool ApplePS2ALPSGlidePoint::deviceSpecificInit() {
     
-    int x = (packet[1] & 0x7f) | ((packet[2] & 0x78) << (7-3));
-    int y = (packet[4] & 0x7f) | ((packet[3] & 0x70) << (7-4));
-    int z = packet[5]; // touch pression
+    // Setup expected packet size
+    priv.pktsize = priv.proto_version == ALPS_PROTO_V4 ? 8 : 6;
     
-    clock_get_uptime(&now_abs);
+    if (!(this->*hw_init)()) {
```

Comment:
For C++ it would have made sense to utilise virtual methods and child classes per touchpad, but I guess we can keep it as is, since this is a port anyway.